### PR TITLE
stats: add more well-known tags

### PR DESCRIPTION
**Pull request by yskopets**
Description: Add more well-known tags
Risk Level: Low
Testing: unit test, manual testing
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: test/common/stats/tag_extractor_impl_test.cc [hunk 1]
```diff
@@ -153,10 +153,34 @@ TEST(TagExtractorTest, DefaultTagExtractors) {
   // Cipher
   Tag cipher_name;
   cipher_name.name_ = tag_names.SSL_CIPHER;
-  cipher_name.value_ = "AES256-SHA";
```
Comment (jmarantz) [inline test/common/stats/tag_extractor_impl_test.cc]: Should we leave this testcase here and just add a new one? Or is it now no longer a test that makes sense?

File: test/common/stats/tag_extractor_impl_test.cc [hunk 2]
```diff
@@ -166,16 +190,28 @@ TEST(TagExtractorTest, DefaultTagExtractors) {
...
+  regex_tester.testRegex("cluster.ratelimit.ssl.sigalgs.rsa_pss_rsae_sha256", "cluster.ssl.sigalgs",
+                         {cluster_tag, ssl_sig_alg});
+
+  // SSL protocol version
+  regex_tester.testRegex("cluster.ratelimit.ssl.versions.TLSv1.2", "cluster.ssl.versions",
+                         {cluster_tag, ssl_version});
+
   // ipv6 non-loopback (for alphabetical chars)
   listener_address.value_ = "[2001_0db8_85a3_0000_0000_8a2e_0370_7334]_3543";
-  regex_tester.testRegex(
```
Comment (jmarantz) [inline test/common/stats/tag_extractor_impl_test.cc]: same with this one. this old one no longer makes sense?

File: test/common/stats/tag_extractor_impl_test.cc [hunk 3]
```diff
@@ -166,16 +190,28 @@ TEST(TagExtractorTest, DefaultTagExtractors) {
...
-      "listener.[2001_0db8_85a3_0000_0000_8a2e_0370_7334]_3543.ssl.cipher.AES256-SHA",
-      "listener.ssl.cipher", {listener_address, cipher_name});
+  regex_tester.testRegex("listener.[2001_0db8_85a3_0000_0000_8a2e_0370_7334]_3543.ssl.ciphers."
+                         "ECDHE-RSA-AES128-GCM-SHA256",
+                         "listener.ssl.ciphers", {listener_address, cipher_name});
 
   // ipv4 address
   listener_address.value_ = "127.0.0.1_0";
-  regex_tester.testRegex("listener.127.0.0.1_0.ssl.cipher.AES256-SHA", "listener.ssl.cipher",
```
Comment (jmarantz) [inline test/common/stats/tag_extractor_impl_test.cc]: ditto

File: test/common/stats/tag_extractor_impl_test.cc [hunk 4]
```diff
@@ -346,6 +382,35 @@ TEST(TagExtractorTest, DefaultTagExtractors) {
...
+
+  regex_tester.testRegex("grpc.sds_mesh_ca.streams_total", "grpc.streams_total",
+                         {google_grpc_service_prefix});
+
+  // GRPC status
+  Tag grpc_status;
+  grpc_status.name_ = tag_names.GRPC_STATUS;
+  grpc_status.value_ = "14";
+
+  regex_tester.testRegex("grpc.sds_mesh_ca.streams_closed_14", "grpc.streams_closed",
```
Comment (jmarantz) [inline test/common/stats/tag_extractor_impl_test.cc]: can you confirm we have testcase covering the 11 new regexes you added

File: test/common/stats/tag_extractor_impl_test.cc (hunk 1 above)
Comment (yskopets) [reply test/common/stats/tag_extractor_impl_test.cc]: As far as I understand from docs, source code and runtime behaviour, this test case is no longer representative.
* According to the [docs](https://github.com/envoyproxy/envoy/blame/master/docs/root/configuration/listeners/stats.rst#L30), the metric is called `ssl.ciphers.<cipher>`
* The [source code](https://github.com/envoyproxy/envoy/blob/master/source/extensions/transport_sockets/tls/context_impl.cc#L70) fills in metric `ssl.ciphers`
* At runtime, Envoy v1.10.0, v1.11.0, v1.12.0, v1.13.0 produce `ssl.ciphers.<cipher>` metric:
```shell
# wget -qO- http://localhost:9901/stats | grep ssl

# notice listener.0.0.0.0_10000.ssl.ciphers.ECDHE-RSA-CHACHA20-POLY1305: 1

cluster.service_google.client_ssl_socket_factory.ssl_context_update_by_sds: 0
cluster.service_google.client_ssl_socket_factory.upstream_context_secrets_not_ready: 0
cluster.service_google.ssl.ciphers.ECDHE-ECDSA-AES128-GCM-SHA256: 1
cluster.service_google.ssl.connection_error: 0
cluster.service_google.ssl.curves.X25519: 1
cluster.service_google.ssl.fail_verify_cert_hash: 0
cluster.service_google.ssl.fail_verify_error: 0
cluster.service_google.ssl.fail_verify_no_cert: 0
cluster.service_google.ssl.fail_verify_san: 0
cluster.service_google.ssl.handshake: 1
cluster.service_google.ssl.no_certificate: 0
cluster.service_google.ssl.session_reused: 0
cluster.service_google.ssl.sigalgs.ecdsa_secp256r1_sha256: 1
cluster.service_google.ssl.versions.TLSv1.2: 1
http.admin.downstream_cx_ssl_active: 0
http.admin.downstream_cx_ssl_total: 0
http.ingress_http.downstream_cx_ssl_active: 0
http.ingress_http.downstream_cx_ssl_total: 1
listener.0.0.0.0_10000.server_ssl_socket_factory.downstream_context_secrets_not_ready: 0
listener.0.0.0.0_10000.server_ssl_socket_factory.ssl_context_update_by_sds: 0
listener.0.0.0.0_10000.server_ssl_socket_factory.upstream_context_secrets_not_ready: 0
listener.0.0.0.0_10000.ssl.ciphers.ECDHE-RSA-CHACHA20-POLY1305: 1
listener.0.0.0.0_10000.ssl.connection_error: 0
listener.0.0.0.0_10000.ssl.curves.X25519: 1
listener.0.0.0.0_10000.ssl.fail_verify_cert_hash: 0
listener.0.0.0.0_10000.ssl.fail_verify_error: 0
listener.0.0.0.0_10000.ssl.fail_verify_no_cert: 0
listener.0.0.0.0_10000.ssl.fail_verify_san: 0
listener.0.0.0.0_10000.ssl.handshake: 1
listener.0.0.0.0_10000.ssl.no_certificate: 1
listener.0.0.0.0_10000.ssl.session_reused: 0
listener.0.0.0.0_10000.ssl.versions.TLSv1.2: 1
```

File: test/common/stats/tag_extractor_impl_test.cc (hunk 2 above)
Comment (yskopets) [reply test/common/stats/tag_extractor_impl_test.cc]: I don't see metric `ssl.cipher.AES256-SHA` generated by any of Envoy v1.10.0, v1.11.0, v1.12.0, v1.13.0

File: test/common/stats/tag_extractor_impl_test.cc (hunk 3 above)
Comment (yskopets) [reply test/common/stats/tag_extractor_impl_test.cc]: I don't see metric `ssl.cipher.AES256-SHA` generated by any of Envoy v1.10.0, v1.11.0, v1.12.0, v1.13.0

File: test/common/stats/tag_extractor_impl_test.cc (hunk 4 above)
Comment (yskopets) [reply test/common/stats/tag_extractor_impl_test.cc]: Yes, I confirm that every new regex has a unit test

**Comment by jmarantz** [comment]
/retest

I'm assuming the coverage failure:
```
[----------] 1 test from Protocols/BufferIntegrationTest
[ RUN      ] Protocols/BufferIntegrationTest.RouteOverride/IPv4_Http2Downstream_HttpUpstream
external/bazel_tools/tools/test/collect_coverage.sh: line 128:  9565 Segmentation fault      (core dumped) "$@"
```
is a legit flake but not related to this PR. Will report that as a separate bug.

File: source/common/config/well_known_names.cc [hunk 5]
```diff
@@ -70,12 +70,33 @@ TagNameValues::TagNameValues() {
   // http.[<stat_prefix>.]fault.(<downstream_cluster>.)<base_stat>
   addRegex(FAULT_DOWNSTREAM_CLUSTER, R"(^http(?=\.).*?\.fault\.((.*?)\.)\w+?$)", ".fault.");
 
-  // listener.[<address>.]ssl.cipher.(<cipher>)
-  addRegex(SSL_CIPHER, R"(^listener(?=\.).*?\.ssl\.cipher(\.(.*?))$)");
+  // listener.[<address>.]ssl.ciphers.(<cipher>)
+  addRegex(SSL_CIPHER, R"(^listener(?=\.).*?\.ssl\.ciphers(\.(.*?))$)", ".ssl.ciphers.");
```
Comment (PiotrSikora) [inline source/common/config/well_known_names.cc]: This probably should be `SSL_CIPHER_SUITE` to match the cluster version.

File: source/common/config/well_known_names.h [hunk 6]
```diff
@@ -82,6 +82,12 @@ class TagNameValues {
   const std::string SSL_CIPHER = "envoy.ssl_cipher";
   // SSL cipher suite
   const std::string SSL_CIPHER_SUITE = "cipher_suite";
```
Comment (PiotrSikora) [inline source/common/config/well_known_names.h]: Should we rename this to add `envoy.` prefix to match other things here?

File: test/common/stats/tag_extractor_impl_test.cc (hunk 1 above)
Comment (PiotrSikora) [reply test/common/stats/tag_extractor_impl_test.cc]: I have no idea where the `ssl.cipher.<cipher>` is coming from. Maybe some mocks produce it?

File: source/common/config/well_known_names.cc (hunk 5 above)
Comment (yskopets) [reply source/common/config/well_known_names.cc]: Removed `SSL_CIPHER` tag completely in favour of `SSL_CIPHER_SUITE`

File: source/common/config/well_known_names.h (hunk 6 above)
Comment (yskopets) [reply source/common/config/well_known_names.h]: I think it's a good idea, but it's a breaking change.

Do you want me to go ahead with the breaking change ?

**Comment by jmarantz** [comment]
maybe merge master and push again?

**Pull request closed without merging**